core/src/main/java/dk/alexandra/fresco/framework/util/HmacDrbg.java

```diff
@@ -0,0 +1,143 @@
+package dk.alexandra.fresco.framework.util;
+
+import dk.alexandra.fresco.framework.MPCException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements the HMac-DRBG (HMac-Deterministic Random Bit Generator) as specified in
+ * https://csrc.nist.gov/publications/detail/sp/800-90a/rev-1/final
+ *
+ * <p>
+ * Note that this class is not threadsafe.
+ * </p>
+ * <p>
+ * Note also that reseeding (calling {@link #setSeed(byte[])}) should occur at least every 2^42
+ * calls to {@link #nextBytes(byte[])} to be secure. No signal will be send about this.
+ * </p>
+ */
+public class HmacDrbg implements Drbg {
+
+  private final Logger logger =
+      LoggerFactory.getLogger(HmacDrbg.class);
+  private static final String DEFAULT_ALGORITHM = "HmacSHA256";
+  private final String algorithm;
+  private Mac mac = null;
+  private byte[] val = null; // value - internally used
+  private byte[] key = null; // key material
+  long reseedCounter;
+  private List<byte[]> seeds;
+  static final long MAX_RESEED_COUNT = (long) Math.pow(2, 48);
+
+  /**
+   * Start with a seed list and the default algorithm. See
+   * {@link #HmacDeterministicRandomBitGeneratorImpl(String, byte[]...)} for further info.
+   *
+   * @param seeds The seeds used. If empty, the default 0 seed will be used. NB: This should
+   *     happen
+   *     only during testing as this is highly insecure.
+   * @throws NoSuchAlgorithmException If the default HmacSHA256 hash function is not found on
+   *     the
+   *     system.
+   */
+  public HmacDrbg(byte[]... seeds) throws NoSuchAlgorithmException {
+    this(DEFAULT_ALGORITHM, seeds);
```

Comment:
Would this class ever work with a different set of algortims? I mean the reseed counter and keys are closely related to the algorithm. Why expose both constructors?

Reply:
You could for instance use the algorithm HmacSHA512. This would work just fine. Just added a test for this. Will commit shortly

Reply:
Well, it does not accept any String, this means we are disclosing a very complicated API within this constructor.

I think the generic case should take a supplier of Mac instead of a String. It would make the interface to the users more complete.